src/PromisifiedFS.js

```diff
@@ -0,0 +1,181 @@
+    this.writeFile = this._wrap(this._writeFile.bind(this))
+    this.unlink = this._wrap(this._unlink.bind(this))
+    this.readdir = this._wrap(this._readdir.bind(this))
+    this.mkdir = this._wrap(this._mkdir.bind(this))
+    this.rmdir = this._wrap(this._rmdir.bind(this))
+    this.rename = this._wrap(this._rename.bind(this))
+    this.stat = this._wrap(this._stat.bind(this))
+    this.lstat = this._wrap(this._lstat.bind(this))
+    // Needed so things don't break if you destructure fs and pass individual functions around
+    this.readFile = this.readFile.bind(this)
+    this.writeFile = this.writeFile.bind(this)
+    this.unlink = this.unlink.bind(this)
+    this.readdir = this.readdir.bind(this)
+    this.mkdir = this.mkdir.bind(this)
+    this.rmdir = this.rmdir.bind(this)
+    this.rename = this.rename.bind(this)
+    this.stat = this.stat.bind(this)
+    this.lstat = this.lstat.bind(this)
+    this.readlink = this.readlink.bind(this)
+    this.symlink = this.symlink.bind(this)
+  }
+  async _init() {
+    if (this._initPromise) return this._initPromise
+    if (this._opts.wipe) {
+      await this._wipe();
+    } else {
+      await this._loadSuperblock();
+    }
+    await this._saveSuperblock();
+  }
+  async _wipe() {
+    await this._idb.wipe()
+    if (this._http) {
+      const text = await this._http.fetchSuperblock()
+      if (text) {
+        this._cache.loadSuperBlock(text)
+      }
+    }
+  }
+  _saveSuperblock() {
+    return this._idb.storeSuperblock(this._cache._root);
+  }
+  async _loadSuperblock() {
+    let root = await this._idb.fetchSuperblock()
+    if (!root && this._http) {
+      root = await this._http.fetchSuperblock()
+    }
+    if (root) {
+      this._cache.loadSuperBlock(root);
+    }
+  }
+  _wrap (fn) {
+    return async (...args) => {
+      await this._init()
+      await this._loadSuperblock()
+      let data, err
+      try {
+        data = await fn(...args)
+      } catch (e) {
+        err = e
+      }
+      await this._saveSuperblock()
+      if (err) throw err
+      return data
+    }
+  }
+  async _readFile(filepath, opts) {
+    ;[filepath, opts] = cleanParams(filepath, opts);
+    const { encoding } = opts;
+    if (encoding && encoding !== 'utf8') throw new Error('Only "utf8" encoding is supported in readFile');
+    const stat = this._cache.stat(filepath);
+    let data = await this._idb.readFile(stat.ino)
+    if (!data && this._http) {
+      data = await this._http.readFile(filepath)
+    }
+    if (data && encoding === "utf8") {
+        data = decode(data);
+    }
+    return data;
+  }
+  async _writeFile(filepath, data, opts) {
+    ;[filepath, opts] = cleanParams(filepath, opts);
+    const { mode, encoding = "utf8" } = opts;
+    if (typeof data === "string") {
+      if (encoding !== "utf8") {
+        throw new Error('Only "utf8" encoding is supported in writeFile');
+      }
+      data = encode(data);
+    }
+    let stat = this._cache.writeFile(filepath, data, { mode });
```

Comment:
Any reason not to have `const` here?

Reply:
Nope. It probably used to live outside of a try/catch. I see a couple other lets I can make consts now.